storage/range.go

```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
 // struct to get args and reply and then applying the command to the
 // state machine via applyRaftCommand(). The error result is sent on
 // the command's done channel, if available.
-//
-// TODO(spencer): Differentiate between errors caused by the normal culprits --
-// bad inputs from clients, stale information, etc. and errors which might
-// cause the range replicas to diverge -- running out of disk space, underlying
-// rocksdb corruption, etc. Do a careful code audit to make sure we identify
-// errors which should be classified as a ReplicaCorruptionError--when those
-// bubble up to the point where we've just tried to execute a Raft command, the
-// Raft replica would need to stall itself.
 func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.InternalRaftCommand) error {
 	if index == 0 {
 		log.Fatalc(r.context(), "processRaftCommand requires a non-zero index")
 	}
+
 	r.Lock()
 	cmd := r.pendingCmds[idKey]
 	delete(r.pendingCmds, idKey)
 	r.Unlock()
 
 	args := raftCmd.Cmd.GetValue().(proto.Request)
+	origin := proto.RaftNodeID(raftCmd.OriginNodeID)
+	var reply proto.Response
+	var ctx context.Context
+	var err error
 	if cmd != nil {
 		// We initiated this command, so use the caller-supplied reply.
-		err := r.applyRaftCommand(cmd.ctx, index, proto.RaftNodeID(raftCmd.OriginNodeID), args, cmd.Reply)
-		cmd.done <- err
-		return err
+		reply = cmd.Reply
+		ctx = cmd.ctx
+	} else {
+		// This command originated elsewhere so we must create a new reply buffer.
+		reply = args.CreateReply()
+		ctx = r.context()
+	}
+
+	// Verify the leader lease is held; Note that we don't require the leader
+	// lease when trying to grant the leader lease!
+	held, expired := verifyLeaderLease(r.getLease(), origin, args.Header().Timestamp)
+	if (!held || expired) && args.Method() != proto.InternalLeaderLease {
+		// This error case is somewhat special: Any command that makes it into
```

Comment:
This is a big comment for a byzantine corruption failure. If the only way this can happen is through a corruption of internal state, then it's simply not something we should be working to handle properly. If we did that for every place in the code where an adversarial internal corruption could occur, we would just end up with an endlessly growing file of special case checks and error handling. If this should never happen, then assert that with a panic here.